### DISCUSSION TITLE
Diarization transcript and normal transcript results differ for the same models & parameters. Is it possible to get the normal transcript with timestamps?

### DISCUSSION THREAD

**Discussion by GreatHalf** [NVIDIA-NeMo/NeMo#5219] [Q&A] [answered]
Hello,

I get different transcript results with the same models and parameters, depending on if I'm using the diarization script or not. I would like to have the exact same transcript process but with timestamps.

### My approach

I followed this tutorial https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/asr_language_modeling.html to customize the language model. It worked, and the results of eval_beamsearch_ngram.py are really good. 
For my case it is important that keywords are written correctly.
Before utilizing the Language Model, the keywords would often get misspelled. Given the domain of the conversations, this is to be expected. With the Language Model however, we get really good results.  I transcribe with the following parameters:

`python3 eval_beamsearch_ngram.py --nemo_model_file …/stt_de_conformer_ctc_large.nemo --kenlm_model_file …/my_model.bin - --input_manifest …/manifest.json --beam_width 128 --beam_alpha 1.0 --beam_beta 0.5 --device cpu --beam_batch_size 1 --acoustic_batch_size 1`

Then I wanted to get per-word-timestamps as well. I do not care about the Speaker Diarization itself, as I know there is only one speaker in the audio. The timestamps are vital however.
It was recommended to use the diarizer to get the timestamps. Luckily, I use a CTC model which is supported.
I tried to keep everything exactly as in the regular transcribing. I'm using the same models and parameters.

`python3 .../examples/speaker_tasks/diarization/clustering_diarizer/offline_diar_with_asr_infer.py diarizer.manifest_filepath=…/manifest.json diarizer.out_dir='diar' diarizer.speaker_embeddings.model_path='titanet_large' diarizer.asr.model_path=…/stt_de_conformer_ctc_large.nemo diarizer.asr.parameters.asr_based_vad=True diarizer.speaker_embeddings.parameters.save_embeddings=False diarizer.asr.ctc_decoder_parameters.pretrained_language_model=…/my_model.bin diarizer.asr.ctc_decoder_parameters.beam_width=128 diarizer.asr.ctc_decoder_parameters.alpha=1.0 diarizer.asr.ctc_decoder_parameters.beta=0.5 diarizer.vad.model_path=vad_multilingual_marblenet diarizer.clustering.parameters.oracle_num_speakers=True diarizer.vad.parameters.filter_speech_first=False diarizer.msdd_model.model_path='diar_msdd_telephonic' `

However, the results are quite different. Unfortunately, the diarizer transcripts are much worse. Once again we are seeing a lot of misspelled words. It's almost as if the language model "my_model.bin" doesn't work anymore.

### Hypotheses

1. You use different CTC python code for the two scripts. Naturally, it will work different in some cases. We have ctc-decoders in https://github.com/NVIDIA/OpenSeq2Seq for the normal transcripts, and pyctcdecode for the diarization transcripts.
2. The diarization transcript does not load the language model correctly. These warnings indicate this. However: Using the arpa version of the language model does not change the results.
```
Unigrams not provided and cannot be automatically determined from LM file (only arpa format). Decoding accuracy might be reduced.
Alphabet determined to be of BPE style.
CTC blank char '' not found, appending to end.
Found <unk> in vocabulary, substituting with ▁⁇▁.
```

In any case, I would like to force the diarization transcription to use the same code as the normal transcription.
Is this possible?

Cheers.

**Comment by tango4j** [comment, marked as answer]
@GreatHalf Unfortunately, `offline_diar_with_asr_infer.py` uses two different systems compared with `eval_beamsearch_ngram.py`.
(1) Buffered conformer CTC: Due to the memory constraint and word timestamp accuracy, speaker diarization uses very short (1~5 sec) buffer length for ASR than stitch the short buffers to render the final output. This could change WER (could be better/worse).
(2) `offline_diar_with_asr_infer.py` uses pyctcdecoder while `eval_beamsearch_ngram.py` uses [open seq2seq ](https://github.com/NVIDIA/OpenSeq2Seq). For speaker diarization, we cannot use open seq2seq since open seq2seq does not support word timestamp outputs. We have tested that if we optimize alpha and beta, we can get equal level of performance with OpenSeq2seq with `pyctcdecode`. However, you need to use **different alpha and beta values that are optimized** for `pyctcdecode` since these two libraries have different implementations of beam search decoder.

Also `pyctcdecode` model is based on word-level LM. Try using [4-gram Big ARPA](https://kaldi-asr.org/models/5/4gram_big.arpa.gz) and provide the model path. The description about the LM models for pyctcdecode is in [README](https://github.com/NVIDIA/NeMo/tree/main/examples/speaker_tasks/diarization) for diarization.
